File: server/src/services/validate.js
const nJ = require('normalize-json');
const got = require('got');

// TODO integrate all this info into the validators
const roomOptionsSchema = nJ({
    title: [String, 30],
    desc: [{ $optional: String }, 255],
});
const addCharaSchema = nJ({
    name: [String, 30],
    color: /^#[0-9a-f]{6}$/g,
    challenge: [String, 128],
});
const addMessageSchema = nJ({
    content: [String, 10000],
    type: ['narrator', 'chara', 'ooc'],
    charaId: msg => (msg.type === 'chara' ? [String] : undefined), // TODO validate chara exists?
    challenge: [String, 128],
});
const editCharaSchema = nJ({
    id: [String],
    name: [String, 30],
    color: /^#[0-9a-f]{6}$/g,
    secret: [String, 64],
});
const editMessageSchema = nJ({
    id: [String],
    content: [String, 10000],
    secret: [String, 64],
});

const validators = {
    async msgs(body) {
        const { type } = body;
        if (type === 'image') {
            const { url = null } = body;
            if (typeof url !== 'string') throw { code: 'BAD_URL' }
            // validate image
            let res;
            try {
                res = await got.head(url);
            } catch (err) {
                throw { code: 'URL_FAILED', details: err.message };
            }
            if (!res.headers['content-type']) throw { code: 'UNKNOWN_CONTENT' };
            if (!res.headers['content-type'].startsWith('image/')) throw { code: 'BAD_IMAGE_CONTENT_TYPE' };
            // ok
            return { type, url };
        }
        else if (type === 'chara') {
            const { content, charaId } = body;
            // TODO validate chara exists?
            return { type, content, charaId };
        }
        else {
            const { content } = body;
            return { type, content };
        }
    },
    async charas(body) {
        const { name, color } = body;
        return { name, color };
    },
    async meta(body) {
        const { title, desc } = body;
        return { title, desc };
    },
}

async function validate(collection, body) {
    if (!validators[collection]) throw new Error('Invalid collection');
    return validators[collection](body);
}

module.exports = validate;

File: server/validate.go
package main

import (
	"fmt"
	"regexp"
)

func (m RpMessageBody) Validate() error {
	if m.Type == "image" {
		if m.Content != "" {
			return fmt.Errorf("Msg: image should not have 'content'")
		}
		if m.CharaID != "" {
			return fmt.Errorf("Msg: image should not have 'charaId'")
		}
		if len(m.URL) > 1000 {
			return fmt.Errorf("Msg: url is too long (%d characters)", len(m.Content))
		}
		urlRegexp := regexp.MustCompile("^https?://[-A-Za-z0-9+&@#/%?=~_|!:,.;]+$")
		if !urlRegexp.MatchString(m.URL) {
			return fmt.Errorf("Msg: url is invalid")
		}
		return nil
	} else if m.Type == "narrator" || m.Type == "chara" || m.Type == "ooc" {
		if m.URL != "" {
			return fmt.Errorf("Msg: non-image should not have 'url'")
		}
		if m.Content == "" {
			return fmt.Errorf("Msg: content is empty")
		}
		if len(m.Content) > 50000 {
			return fmt.Errorf("Msg: content is too long (%d characters)", len(m.Content))
		}
		if m.Type == "chara" {
			if m.CharaID == "" {
				return fmt.Errorf("Msg: charaId is empty")
			}
		} else {
			if m.CharaID != "" {
				return fmt.Errorf("Msg: non-chara msg should not have 'charaId'")
			}
		}
		return nil
	} else {
		return fmt.Errorf("Msg: invalid type")
	}
}

func (m RpMessageBody) CheckRelations(rpid string, db *database) error {
	if m.Type == "chara" {
		if db.getChara(rpid, m.CharaID) == nil {
			return fmt.Errorf("Msg with invalid chara id: %s", m.CharaID)
		}
	}
	return nil
}

func (c RpCharaBody) Validate() error {
	if len(c.Name) == 0 {
		return fmt.Errorf("Chara: name is empty")
	}
	if len(c.Name) > 30 {
		return fmt.Errorf("Chara: name is too long")
	}
	colorRegexp := regexp.MustCompile("^#[0-9a-f]{6}$")
	if !colorRegexp.MatchString(c.Color) {
		return fmt.Errorf("Chara: color is invalid")
	}
	return nil
}

func (c RpCharaBody) CheckRelations(rpid string, db *database) error {
	return nil
}
